client/src/main/java/io/pravega/client/stream/impl/ControllerResolverFactory.java

```diff
@@ -161,26 +158,60 @@ public void shutdown() {
             }
         }
 
+        @Override
+        @Synchronized
+        public void refresh() {
+            if (started.get()) {
+                this.scheduledExecutor.execute(this::getControllers);
+            }
+        }
+
         // The controller discovery API invoker.
         private void getControllers() {
+            log.debug("Attempting to refresh the controller server endpoints");
             try {
-                final ServerResponse controllerServerList =
-                        this.client.getControllerServerList(ServerRequest.getDefaultInstance());
-                final ResolvedServerInfoGroup serverInfoGroup = ResolvedServerInfoGroup.builder()
-                        .addAll(controllerServerList.getNodeURIList()
-                                .stream()
-                                .map(node ->
-                                        new ResolvedServerInfo(
-                                                new InetSocketAddress(node.getEndpoint(), node.getPort())))
-                                .collect(Collectors.toList()))
-                        .build();
+                final ResolvedServerInfoGroup serverInfoGroup;
+                if (this.enableDiscovery) {
+                    final ServerResponse controllerServerList =
+                            this.client.getControllerServerList(ServerRequest.getDefaultInstance());
+                    serverInfoGroup = ResolvedServerInfoGroup.builder()
+                            .addAll(controllerServerList.getNodeURIList()
+                                    .stream()
+                                    .map(node ->
+                                            new ResolvedServerInfo(
+                                                    new InetSocketAddress(node.getEndpoint(), node.getPort())))
+                                    .collect(Collectors.toList()))
+                            .build();
+                } else {
+                    // Resolve the bootstrapped server list to get the set of controllers.
+                    final ArrayList<InetSocketAddress> resolvedAddresses = new ArrayList<>();
+                    this.bootstrapServers.forEach(address -> {
+                        try {
+                            resolvedAddresses.add(new InetSocketAddress(InetAddress.getByName(address.getHostString()),
```

Comment:
On creating a instance of InetSocketAddress it tries to resolve it. It will flag it as unresolved if it is not able to lookup. `java.net.InetSocketAddress#isUnresolved` returns true. If unresolved we can skip adding it.